### PR TITLE
app-layer: websockets protocol support

### PR DESCRIPTION
**Pull request by catenacyber**
Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket:
https://redmine.openinfosecfoundation.org/issues/2695

Describe changes:
- app-layer: websockets protocol support

Just a draft to show started work

```
SV_BRANCH=pr/1519
```
https://github.com/OISF/suricata-verify/pull/1519

#9987 with more TODOs done

What is left todo :
- keywords
- reassembly ?

### CODE REVIEW

**Comment by suricata-qa** [comment]
WARNING:
 
 | field | baseline | test | % |
|--|--|--|--| 
| | SURI_TLPR1_stats_chk | | |
| .tcp.pseudo | 2810 | 3014 | 107.26% | 

Pipeline 16907

**Comment by catenacyber** [comment]
Also https://github.com/OISF/suricata/pull/9870 should be merged before this to avoid a short-lived output-json-websockets.c file ;-)

File: etc/schema.json [hunk 1]
```diff
@@ -3823,6 +3823,9 @@
                                 },
                                 "tls": {
                                     "$ref": "#/$defs/stats_applayer_error"
+                                },
+                                "websockets": {
```
Comment (victorjulien) [inline etc/schema.json:3827]: rfc 6455 refers to it as "WebSocket" (singular), so lets use singular in our code and docs

File: rust/src/websockets/logger.rs [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+        8 => Some("connection_close"),
+        9 => Some("ping"),
+        0xa => Some("pong"),
+        _ => None,
+    }
+}
+
+fn log_websockets(tx: &WebSocketsTransaction, js: &mut JsonBuilder) -> Result<(), JsonError> {
+    js.open_object("websockets")?;
+    js.set_bool("mask", tx.pdu.mask)?;
```
Comment (victorjulien) [inline rust/src/websockets/logger.rs:38]: would it make sense to log the mask value?

File: rust/src/websockets/websockets.rs [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+                        WebSocketsFrameType::Pdu as u8,
+                    );
+                    start = rem;
+                    let mut tx = self.new_tx(direction);
+                    tx.pdu = pdu;
+                    //TODOws should we reassemble/stream payload data ?
+                    self.transactions.push_back(tx);
+                }
+                Err(nom::Err::Incomplete(_)) => {
+                    // Not enough data. just ask for one more byte.
```
Comment (victorjulien) [inline rust/src/websockets/websockets.rs:139]: as the header seems to have a payload_len, could we use that here?

Also what is the max payload size?

File: src/app-layer-htp.c [hunk 4]
```diff
@@ -975,29 +976,52 @@ static AppLayerResult HTPHandleResponseData(Flow *f, void *htp_state, AppLayerPa
...
+                            SCReturnStruct(APP_LAYER_INCOMPLETE(consumed, input_len - consumed));
+                        }
+                        SCReturnStruct(APP_LAYER_OK);
+                    } else if (bstr_cmp_c_nocase(h->value, "WebSocket") == 0) {
+                        if (AppLayerProtoDetectGetProtoName(ALPROTO_WEBSOCKETS) == NULL) {
+                            // if WS is disabled, keep the HTP_STREAM_TUNNEL mode
+                            break;
+                        }
+                        hstate->slice = NULL;
+                        if (AppLayerExpectationCreate(f, STREAM_TOCLIENT | STREAM_TOSERVER, f->sp,
```
Comment (victorjulien) [inline src/app-layer-htp.c:1009]: surprised to see this. Expectations are about related but unique flows (so different 5 tuple), that is not the case here, right?

**Review by victorjulien — CHANGES_REQUESTED** [review]
Nice work. Left some comments inline.

File: rust/src/websockets/websockets.rs (hunk 3 above)
Comment (jasonish) [reply rust/src/websockets/websockets.rs:139]: A frame, per spec can be up to 9,223,372,036,854,775,807 bytes I believe.

Comment (victorjulien) [reply rust/src/websockets/websockets.rs:139]: but I guess it means we need to consume all payload data directly, even if incomplete. Process them the same as other streaming buffers like http body and file.data

Comment (jasonish) [reply rust/src/websockets/websockets.rs:139]: > frames the size of the universe, of course

In practice, most implementations don't allow extremely large payloads, I think 16kb even being on the extreme..  But we of course still need to deal with universe sized frames somehow.

File: etc/schema.json (hunk 1 above)
Comment (catenacyber) [reply etc/schema.json:3827]: indeed

File: rust/src/websockets/logger.rs (hunk 2 above)
Comment (catenacyber) [reply rust/src/websockets/logger.rs:38]: I do not think so.
Frame can be used to detect on it...

File: src/app-layer-htp.c [hunk 5]
```diff
@@ -975,29 +976,52 @@ static AppLayerResult HTPHandleResponseData(Flow *f, void *htp_state, AppLayerPa
...
+                                    HTTP_DECODER_EVENT_FAILED_PROTOCOL_CHANGE);
+                        }
+                        // During HTTP2 upgrade, we may consume the HTTP1 part of the data
+                        // and we need to parser the remaining part with HTTP2
+                        if (consumed > 0 && consumed < input_len) {
+                            SCReturnStruct(APP_LAYER_INCOMPLETE(consumed, input_len - consumed));
+                        }
+                        SCReturnStruct(APP_LAYER_OK);
+                    } else if (bstr_cmp_c_nocase(h->value, "WebSocket") == 0) {
+                        if (AppLayerProtoDetectGetProtoName(ALPROTO_WEBSOCKETS) == NULL) {
```
[the remark this answers is not in the document]
Comment (victorjulien) [reply src/app-layer-htp.c:1004]: `AppLayerProtoDetectGetProtoName` does a table lookup, but the answer will always be the same. So we could look it up once and store the result

File: src/app-layer-htp.c (hunk 4 above)
Comment (victorjulien) [reply src/app-layer-htp.c:1009]: Expectations create a new flow, so shouldn't be used here. It's the wrong API for the problem.

Forcing the protocol change, perhaps with some validation that it indeed looks like websock, sounds better.

**Comment by catenacyber** [comment]
Replaced by https://github.com/OISF/suricata/pull/10014

File: src/app-layer-htp.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (catenacyber) [reply src/app-layer-htp.c]: Oh. Why is this "table lookup" fairly expensive ? 
To me, it looks negligible next to the string comparison

File: src/app-layer-htp.c (hunk 4 above)
Comment (catenacyber) [reply src/app-layer-htp.c:1009]: Now, that looks evil 👿 
Thanks for the explanation :-)

Comment (catenacyber) [reply src/app-layer-htp.c:1009]: Got a new hack for next PR iteration

**Pull request closed without merging**